[feat] it is a web app that will (create , read , update, delete) pro…

MhmudAbdGhafar opened the pull request:
…ducts of a specific e-commerce app using:

 spring mvc for backend ,
hibernate for database operations
MySQL for data store
 maven for dependency management
Bootstrap , HTML and CSS for frontend || #15 | SimpleMonsterWar

Commit:
[feat] it is a web app that will (create , read , update, delete) products of a specific e-commerce app using:
 spring mvc for backend ,
hibernate for database operations
MySQL for data store
 maven for dependency management
Bootstrap , HTML and CSS for frontend || #15 | SimpleMonsterWar

Commit:
[feat] add backend validation ,jsp files, dao layer, service layer, controller requests || #15 | SimpleMonsterWar

salma-4 commented on SimpleMonsterWar/src/main/java/com/ecommerce/app/model/ProductDetails.java:
```diff
@@ -0,0 +1,56 @@
+
+    @NotBlank(message = "Name is required")
+    @Column(name = "name")
+    private String name;
```
no need for another name attribute as product table already has name

salma-4 commented on SimpleMonsterWar/src/main/java/com/ecommerce/app/controller/ProductController.java:
```diff
@@ -0,0 +1,102 @@
+    @PostMapping("/save")
+    public String saveProduct(
+            @Valid @ModelAttribute("product") Product product,
+            @ModelAttribute("details") ProductDetails details,
```
Since the Product entity already contains a **ProductDetails object**, there is no need to pass ProductDetails as a separate model attribute. Spring will automatically bind all nested fields (e.g., productDetails.name, productDetails.price, etc.) into the Product object during form submission

salma-4 commented on SimpleMonsterWar/src/main/java/com/ecommerce/app/dao/ProductDAOImpl.java:
```diff
@@ -0,0 +1,51 @@
+
+        Session currentSession = sessionFactory.getCurrentSession();
+
+        currentSession.persist(product);
```
**persist()** is intended only for inserting new records. Using it during an update operation will not modify the existing row — instead, Hibernate may attempt to treat it as a new entity. To correctly handle both insert and update scenarios, it's better to use **saveOrUpdate()**, which automatically decides whether to create a new record or update an existing one based on the entity’s ID.

salma-4 commented on SimpleMonsterWar/src/main/java/com/ecommerce/app/model/Product.java:
```diff
@@ -0,0 +1,36 @@
+    @Column(name = "name")
+    private String name;
+
+    @OneToOne(cascade = CascadeType.ALL)
```
CascadeType.ALL on OneToOne is  **dangerous** :
- If you delete product → details deleted (okay).
- But if you update product → Hibernate may delete & re-insert details.
so you can **CascadeType.Persist**

salma-4 commented on SimpleMonsterWar/src/main/webapp/WEB-INF/application-context.xml:
```diff
@@ -0,0 +1,62 @@
+        <property name="packagesToScan" value="com.ecommerce.app" />
+        <property name="hibernateProperties">
+            <props>
+                <prop key="hibernate.dialect">org.hibernate.dialect.MySQL5Dialect</prop>
```
improvement : org.hibernate.dialect.MySQL8Dialect

Commit:
[feat] it is a web app that will (create , read , update, delete) products of a specific e-commerce app using:

 spring mvc for backend ,
hibernate for database operations
MySQL for data store
 maven for dependency management
Bootstrap , HTML and CSS for frontend || #15 | ProductManagementSystem

Commit:
[fix] remove old unneeded code || #15 | SimpleMonsterWar

Commit:
[fix] fix the issue of fetch lazy and form validation || #15 || ProductManagementSystem

salma-4 commented on ProductManagementSystem/src/main/java/com/ecommerce/app/controller/ProductController.java:
```diff
@@ -0,0 +1,123 @@
+            return "products/add";
+        }
+
+        if (product.getProductDetails() != null) {
```
It is better to make any logic in service instead of controller

salma-4 commented on ProductManagementSystem/src/main/java/com/ecommerce/app/controller/ProductController.java:
```diff
@@ -0,0 +1,123 @@
+
+        if (details == null) {
+            details = new ProductDetails();
+            product.setProductDetails(details);
```
It is better to make any logic in service instead of controller

salma-4 commented on ProductManagementSystem/src/main/java/com/ecommerce/app/service/ProductServiceImpl.java:
```diff
@@ -0,0 +1,63 @@
+
+        Hibernate.initialize(product.getProductDetails());
+
+        return productDAO.findById(id);
```
You called findById two times you can only return product ,

salma-4 reviewed (commented):
Great improvement overall 👏

- Clean code: Clear layering and readable structure make the code easy to understand and maintain
- Best Practices: Java-based configuration, validation, and service-level transactions 
- Performance: Lazy loading and transactional services are well applied

Keep going

Commit:
Merge branch 'main' into session15-simple-project

Commit:
[fix] Document api using swagger and resolve some comments || #15 | ProductManagementSystem

Commit:
[fix] add dto layer to carry product data between controller and dao || #15 | ProductManagementSystem

Commit:
[fix] update existing product in database instead of creating new record in ProductDetails table || #15 | ProductManagementSystem
